File: src/cli/extrusion_painting.py
import sys
from PIL import Image
import numpy as np

def load_image(image_file):
    image = Image.open(image_file).convert('L')
    return np.array(image)

def load_gcode(gcode_file):
    with open(gcode_file, 'r') as file:
        return file.readlines()

def save_gcode(gcode_lines, output_file):
    with open(output_file, 'w') as file:
        file.writelines(gcode_lines)

def modify_extrusion_rates(gcode_lines, image_data, projection_mode, z_offset, min_extrusion, max_extrusion):
    for i, line in enumerate(gcode_lines):
        if line.startswith('G1'):
            intensity = get_intensity(line, image_data, projection_mode, z_offset)
            extrusion_rate = min_extrusion + (max_extrusion - min_extrusion) * intensity
            gcode_lines[i] = adjust_extrusion_in_line(line, extrusion_rate)
    return gcode_lines

def get_intensity(line, image_data, projection_mode, z_offset):
    # Implement the logic to get the intensity based on the projection mode and image data
    # This is a placeholder implementation
    return 0.5

def adjust_extrusion_in_line(line, extrusion_rate):
    parts = line.split()
    for i, part in enumerate(parts):
        if part.startswith('E'):
            parts[i] = f'E{extrusion_rate}'
            break
    return ' '.join(parts) + '\n'

def main():
    if len(sys.argv) != 6:
        print("Usage: python extrusion_painting.py <projection_mode> <params> <image_file> <source_gcode> <output_gcode>")
        sys.exit(1)

    projection_mode = sys.argv[1]
    params = sys.argv[2]
    image_file = sys.argv[3]
    source_gcode = sys.argv[4]
    output_gcode = sys.argv[5]

    image_data = load_image(image_file)
    gcode_lines = load_gcode(source_gcode)

    min_extrusion = 0.5  # Example value, adjust as needed
    max_extrusion = 1.5  # Example value, adjust as needed
    z_offset = 0  # Example value, adjust as needed

    adjusted_gcode = modify_extrusion_rates(gcode_lines, image_data, projection_mode, z_offset, min_extrusion, max_extrusion)
    save_gcode(adjusted_gcode, output_gcode)

if __name__ == "__main__":
    main()

File: src/libslic3r/ExtrusionPainting.cpp
#include "ExtrusionPainting.hpp"
#include <cmath>
#include <vector>
#include <algorithm>

namespace Slic3r {

void ExtrusionPainting::modify_extrusion_rates(std::vector<GCodeLine>& gcode_lines, const std::vector<std::vector<int>>& image_data, int projection_mode, double z_offset, double min_extrusion, double max_extrusion) {
    for (auto& line : gcode_lines) {
        double intensity = get_intensity(line, image_data, projection_mode, z_offset);
        double extrusion_rate = min_extrusion + (max_extrusion - min_extrusion) * intensity;
        line.set_extrusion_rate(extrusion_rate);
    }
}

double ExtrusionPainting::get_intensity(const GCodeLine& line, const std::vector<std::vector<int>>& image_data, int projection_mode, double z_offset) {
    // Implement the logic to get the intensity based on the projection mode and image data
    // This is a placeholder implementation
    return 0.5;
}

void ExtrusionPainting::smooth_transitions(std::vector<GCodeLine>& gcode_lines) {
    // Implement the logic to smooth transitions (anti-aliasing) for gradients
    // This is a placeholder implementation
}

void ExtrusionPainting::process_gcode(std::vector<GCodeLine>& gcode_lines, const std::vector<std::vector<int>>& image_data, int projection_mode, double z_offset, double min_extrusion, double max_extrusion) {
    modify_extrusion_rates(gcode_lines, image_data, projection_mode, z_offset, min_extrusion, max_extrusion);
    smooth_transitions(gcode_lines);
}

} // namespace Slic3r
